tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,405 @@
+  """ Checks that each of tensors has the same backend, returning True and an
+      empty string if so, or False and an error string if not.
+  Args:
+    tensors: The list of tensors whose backends to check.
+    fname: The name of the calling function, which will go into the errstring.
+  Returns:
+    (flag, errstr): Whether all backends agree, and an error message if not.
+  """
+  backend = tensors[0].backend
+  backend_names = [tensor.backend.name for tensor in tensors]
+  all_backends_same = True
+  for name in backend_names[1:]:
+    all_backends_same = all_backends_same and name == backend.name
+  errstr = ""
+  if not all_backends_same:
+    errstr = "All Tensors fed to " + fname + "must have the same backend."
+    errstr += "Backends were: \n"
+    errstr += str([name + "\n" for name in backend_names])
+  return all_backends_same, errstr
+
+
+def tensordot(a: Tensor, b: Tensor, axes: Sequence[Sequence[int]]) -> Tensor:
+  """Do a tensordot (contraction) of Tensors `a` and `b` over the given axes.
+  The behaviour of this function largely matches that of np.tensordot.
+
+  Args:
+    a: A Tensor.
+    b: Another Tensor.
+    axes: Two lists of integers. These values are the contraction
+          axes.
+  Raises:
+    ValueError, if a and b have different backends.
+  Returns:
+    The result of the tensordot, a Tensor.
+  """
+  if a.backend.name != b.backend.name:
+    errstr = "Tried to Tensordot Tensors with differing backends \n"
+    errstr += a.backend.name + "and " + b.backend.name + "."
+    raise ValueError(errstr)
+  out_array = a.backend.tensordot(a.array, b.array, axes)
+  out_tensor = Tensor(out_array, backend=a.backend)
+  return out_tensor
+
+
+def reshape(tensor: Tensor, new_shape: Sequence[int]) -> Tensor:
+  """Reshape Tensor to the given shape.
+
+  Args:
+    tensor: Tensor to reshape.
+    new_shape: The new shape.
+  Returns:
+    The reshaped Tensor.
+  """
+  return tensor.reshape(new_shape)
+
+
+def transpose(tensor: Tensor, perm: Optional[Sequence[int]] = None) -> Tensor:
+  """ Return a new `Tensor` transposed according to the permutation set
+  by `axes`. By default the axes are reversed.
+  Args:
+    axes: The permutation. If None (default) the index order is reversed.
+  Returns:
+    The transposed `Tensor`.
+  """
+  return tensor.transpose(perm=perm)
+
+
+def take_slice(tensor: Tensor, start_indices: Tuple[int, ...],
+               slice_sizes: Tuple[int, ...]) -> Tensor:
+  """Obtains a slice of a Tensor based on start_indices and slice_sizes.
+
+  Args:
+    Tensor: A Tensor.
+    start_indices: Tuple of integers denoting start indices of slice.
+    slice_sizes: Tuple of integers denoting size of slice along each axis.
+  Returns:
+    The slice, a Tensor.
+  """
+  sliced = tensor.backend.slice(tensor.array, start_indices, slice_sizes)
+  sliced_tensor = Tensor(sliced, backend=tensor.backend)
+  return sliced_tensor
+
+
+def shape(tensor: Tensor) -> Tuple[Optional[int], ...]:
```

Comment:
why is that Optional?

Reply:
I dunno what I was thinking here. I changed it.